src/MessageWindow.js

```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import MessageGroup from './MessageGroup';
+import MessageText from './MessageText';
+import { MessageContainerType } from './MessengerTypes';
+import TypingIndicator from './TypingIndicator';
+
+class MessageWindow extends React.Component {
+    render() {
+        let groups = this.props.messagesGroup.map((group, index) => {
```

Comment:
Ogólnie dla kodzenia w JSie, staramy się unikać `let`ów, kod powinien być immutable, bo mutacje stanów i obiektów prowadzą do błędów ciężkich do wykrycia.
W tym przypadku, zamiast modyfikować `groups` w ifie, można to załatwić w samym returnie:
```
return (
  <div>
    {groups}
    {this.props.indicator.isTyping && this.props.indicator.user !== this.props.user && <MessageText key={0} slot={<TypingIndicator />} />}
  </div>
);